Game of Life game application, Iteration 2

EVIJA-OZO opened the pull request:
1.Added information to represent count of iteration and count of live cells.
2. Added possibility to save information to file and restore it.
3. Possibility to pause application in any moment.
4.Added interfaces.

EVIJA-OZO committed:
1.Added information to represent count of iteration and count of live cells.
2. Added possibility to save information to file and restore it.
3. Possibility to pause application in any moment.
4.Added interfaces.

ArvisZalitis commented on GameOfLife/Controller/GameManager.cs:
```diff
@@ -1,29 +1,70 @@
+                {
+                    Console.WriteLine(GameMessages.stopGame);
+                    string choice = Console.ReadLine();
+                    if (choice == "s")
```
If I enter S for example, game wont be saved, you can use here `ToLower` and the same for other choices.

EVIJA-OZO replied:
Changes have been made.

ArvisZalitis commented on GameOfLife/Controller/GameManager.cs:
```diff
@@ -1,29 +1,70 @@
+        /// </summary>
+        public static Game? LoadGame()
+        {
+            Game loadgame = JsonConvert.DeserializeObject<Game>(File.ReadAllText(GameMessages.fileName));
```
No need for `loadgame` variable, simply return.

EVIJA-OZO replied:
If I use simply return, it gives CS0126 error.

ArvisZalitis commented on GameOfLife/Model/Game.cs:
```diff
@@ -1,53 +1,66 @@
+        public int [,] GameField { get; set; }
+
+        public int CountOfIteration { get; set; }
+        public int CountOfLiveCells { get; set; }
```
Missing comments.

EVIJA-OZO replied:
Comments added.

ArvisZalitis reviewed (approved):
Look ok, some small fixes needed but in overall thumb up.

EVIJA-OZO committed:
Added comments in Game.cs
ToLower() added for user input choices.

GuntarsD commented on GameOfLife/View/GameView.cs:
```diff
@@ -9,17 +9,17 @@ public class GameView
         /// </summary>
         /// <param name="args">Array containing the game field.</param>
-        public void View(int[,] args)
+        public static void View(int[,] args)
```
View is not good name for this method.

Probably better:
DisplayGame,
or RenderGame


Anyway make name more descriptive.

EVIJA-OZO replied:
Class and method renamed.

GuntarsD commented on GameOfLife/View/GameView.cs:
```diff
@@ -9,17 +9,17 @@ public class GameView
-                    Console.Write(args[column , row] == 1 ? "*" : " ");
+                    Console.Write(args[column, row] == 1 ? "*" : " ");
                 }
                 Console.WriteLine();
```
Always insert empty line after closing } if that is not followed  by closing } of inclosing block

EVIJA-OZO replied:
Empty lines added.

GuntarsD commented on GameOfLife/Controller/GameManager.cs:
```diff
@@ -1,29 +1,70 @@
+                GameView.View(game.GameField);
+                Console.WriteLine($"Iteration: {game.CountOfIteration}");
+                Console.WriteLine($"Live cells: {game.CountOfLiveCells}");
+                Console.WriteLine(GameMessages.pauseGame);
```
It would be better not to use Console object outside UserInterface class.

It looks that displaying messages should be part of GameView class, because it is part of displaying data on screen.

EVIJA-OZO replied:
26.-28. line moved to DisplayGame.

GuntarsD commented on GameOfLife/Controller/GameManager.cs:
```diff
@@ -1,29 +1,70 @@
+        /// Saves current game generation to the file.
+        /// </summary>
+        /// <param name="game">Current game generation.</param>
+        public static void SaveGame(Game game)
```
It is better to put SaveGame and LoadGame into dedicated class like:
DataManager.
That because file and data management is somewhat different from game management.

EVIJA-OZO replied:
New class DataManager added.

GuntarsD commented on GameOfLife/Model/Game.cs:
```diff
@@ -1,53 +1,71 @@
         /// </summary>
-        public int [,] gameField;
+        public int[,] GameField { get; set; }
+        /// <summary>
```
Here and in other places please insert empty line before comments.

EVIJA-OZO replied:
Changes have been made.

GuntarsD commented on GameOfLife/Model/Game.cs:
```diff
@@ -1,53 +1,71 @@
-                for (int row = 0; row < gameField.GetLength(1); row++)
+                for (int row = 0; row < GameField.GetLength(1); row++)
                 {
                     newGameField[column, row] = AliveCells(column, row);
```
AliveCells is misleading name.

May be better WillCellBeAlive
or similar

EVIJA-OZO replied:
Method is renamed.

GuntarsD commented on GameOfLife/Model/Game.cs:
```diff
@@ -1,53 +1,71 @@
                 }
             }
-            gameField = newGameField;
+            GameField = newGameField;
```
Add empty line

EVIJA-OZO replied:
Changes have been made.

GuntarsD commented on GameOfLife/Model/Game.cs:
```diff
@@ -74,6 +92,7 @@ private int AliveCells(int column, int row)
                 return 0;
```
Replace
else
            {
                return 0;
            }
            
with
    return 0;

EVIJA-OZO replied:
Changes have been made.

EVIJA-OZO committed:
GameView renamed to DisplayGame;
Added empty lines before comments, after closing brackets;
Displaying game data moved to DisplayGame class;
New class DataManager for Save un Load functions;
AliveCells renamed to WillCellBeAlive;

GuntarsD commented on GameOfLife/Controller/DataManager.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace GameOfLife
+{
+    public class DataManager
```
Please add comment to class

EVIJA-OZO replied:
Comment added.

GuntarsD commented on GameOfLife/Controller/DataManager.cs:
```diff
@@ -0,0 +1,26 @@
+        public static Game? LoadGame()
+        {
+            Game loadgame = JsonConvert.DeserializeObject<Game>(File.ReadAllText(GameMessages.fileName));
+            return loadgame;
```
Add empty line before return statement and remove one after the return statement.

EVIJA-OZO replied:
Changes have been made.

EVIJA-OZO committed:
Changes in DataManager - added comment, empty line.
